src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComVariantMarshaller.cs

```diff
@@ -0,0 +1,332 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace System.Runtime.InteropServices.Marshalling
+{
+    /// <summary>
+    /// Marshals an <see cref="object"/> to an <see cref="ComVariant"/>.
+    /// </summary>
+    /// <remarks>
+    /// Supports the same types as <see cref="ComVariant.Create{T}(T)"/> as well as any types with <see cref="GeneratedComClassAttribute"/> applied.
+    /// </remarks>
+    [CustomMarshaller(typeof(object), MarshalMode.Default, typeof(ComVariantMarshaller))]
+    [CustomMarshaller(typeof(object), MarshalMode.UnmanagedToManagedRef, typeof(RefPropagate))]
+    public static partial class ComVariantMarshaller
+    {
+        public static ComVariant ConvertToUnmanaged(object? managed)
+        {
+            if (managed is null)
+            {
+                return default;
+            }
+
+#pragma warning disable CS0618 // Type or member is obsolete
+            switch (managed)
+            {
+                case sbyte s:
+                    return ComVariant.Create(s);
+                case byte b:
+                    return ComVariant.Create(b);
+                case short s:
+                    return ComVariant.Create(s);
+                case ushort s:
+                    return ComVariant.Create(s);
+                case int i:
+                    return ComVariant.Create(i);
+                case uint i:
+                    return ComVariant.Create(i);
+                case long l:
+                    return ComVariant.Create(l);
+                case ulong l:
+                    return ComVariant.Create(l);
+                case float f:
+                    return ComVariant.Create(f);
+                case double d:
+                    return ComVariant.Create(d);
+                case decimal d:
+                    return ComVariant.Create(d);
+                case bool b:
+                    return ComVariant.Create(b);
+                case char c:
+                    return ComVariant.Create((ushort)c);
```

Comment:
Is this supported in some way already? What is the user intent here?

Reply:
The built-in VARIANT marshalling always treats `char` values as `VT_UI2`.

As this is a little unintuitive, I can remove this for V1 and we can add it based on user feedback.